src/dag.c

```diff
@@ -495,48 +661,207 @@ int RedisAI_DagRunSyntaxParser(RedisModuleCtx *ctx, RedisModuleString **argv,
+    AI_dictEntry *entry = AI_dictNext(iter);
+    while (entry) {
+      char *key = (char *)AI_dictGetKey(entry);
+      char *demangled_key = RedisModule_Strdup(key);
+      demangled_key[strlen(key) - 4] = 0;
+      int *instance = RedisModule_Alloc(sizeof(int));
+      *instance = 1;
+      AI_dictAdd(mangled_tensors, (void *)demangled_key, (void *)instance);
+      RedisModule_Free(demangled_key);
+      entry = AI_dictNext(iter);
+    }
+    AI_dictReleaseIterator(iter);
+  }
+
+  for (long long i=0; i<array_len(rinfo->dagOps); i++) {
+    RAI_DagOp *currentOp = rinfo->dagOps[i];
+
+    RedisModuleString **mangled_inkeys = array_new(RedisModuleString*, array_len(currentOp->inkeys));
+    for (long long j=0; j<array_len(currentOp->inkeys); j++) {
+      const char* key = RedisModule_StringPtrLen(currentOp->inkeys[j], NULL);
+      AI_dictEntry *entry = AI_dictFind(mangled_tensors, key);
+      if (!entry) {
+        return RedisModule_ReplyWithError(ctx,
+                                          "ERR INPUT key cannot be found in DAG");
+      }
+      int *instance = AI_dictGetVal(entry);
+      RedisModuleString *mangled_key = RedisModule_CreateStringPrintf(ctx, "%s%04d", key, *instance);
+      mangled_inkeys = array_append(mangled_inkeys, mangled_key);
+    }
+
+    RedisModuleString **mangled_outkeys = array_new(RedisModuleString*, array_len(currentOp->outkeys));
+    for (long long j=0; j<array_len(currentOp->outkeys); j++) {
+      const char* key = RedisModule_StringPtrLen(currentOp->outkeys[j], NULL);
+      AI_dictEntry *entry = AI_dictFind(mangled_tensors, key);
+      int *instance = NULL;
+      if (entry) {
+        instance = AI_dictGetVal(entry);
+        *instance += 1;
+      }
+      else {
+        instance = RedisModule_Alloc(sizeof(int));
+        *instance = 1;
+        AI_dictAdd(mangled_tensors, (void *)key, (void *)instance);
+      }
+      RedisModuleString *mangled_key = RedisModule_CreateStringPrintf(ctx, "%s%04d", key, *instance);
+      mangled_outkeys = array_append(mangled_outkeys, mangled_key);
+    }
+  
+    array_free(currentOp->inkeys);
+    array_free(currentOp->outkeys);
+
+    currentOp->inkeys = mangled_inkeys;
+    currentOp->outkeys = mangled_outkeys;
+  }
+
+  AI_dict* mangled_persisted = AI_dictCreate(&AI_dictTypeHeapStrings, NULL);
+  {
+    AI_dictIterator *iter = AI_dictGetSafeIterator(rinfo->dagTensorsPersistedContext);
+    AI_dictEntry *entry = AI_dictNext(iter);
+    while (entry) {
+      char *key = (char *)AI_dictGetKey(entry);
+      AI_dictEntry *mangled_entry = AI_dictFind(mangled_tensors, key);
+      if (!mangled_entry) {
+        return RedisModule_ReplyWithError(ctx,
+                                          "ERR PERSIST key cannot be found in DAG");
+      } 
+      int *instance = AI_dictGetVal(mangled_entry);
+      RedisModuleString *mangled_key = RedisModule_CreateStringPrintf(ctx, "%s%04d", key, *instance);
```

Comment:
possible leak?

Reply:
Nice catch, now that we mangle we need to free the `inkeys` and `outkeys` strings when freeing the ops.